src/App.tsx

```diff
@@ -1,24 +1,317 @@
+    const deletedTodo = todos?.find(todo => todo.id === targetId);
+
+    try {
+      setTargetTodosIds(currIds => {
+        return [...currIds, deletedTodo?.id || 0];
+      });
+      await deleteTodo(deletedTodo?.id || 0);
+      setTodos(todos?.filter(todo => todo.id !== +targetId) || null);
+    } catch (error) {
+      setTypeOfError(Errors.Deleting);
+    } finally {
+      setTargetTodosIds((currIds: number[]) => {
+        return currIds.filter((id: number) => id !== targetId);
+      });
+    }
+  };
+
+  const clearCompleted = async () => {
+    const todosForDeleting: Promise<unknown>[] = [];
+    const completedTodoIds: number[] = [];
+
+    todos?.forEach(todo => {
+      if (todo.completed) {
+        todosForDeleting.push(deleteTodo(todo.id));
+        completedTodoIds.push(todo.id);
+      }
+    });
+
+    try {
+      setTargetTodosIds(currIds => currIds.concat(completedTodoIds));
+      await Promise.all(todosForDeleting);
+
+      setTodos(todos?.filter(todo => !todo.completed) || null);
+    } catch (error) {
+      setTypeOfError(Errors.Deleting);
+    } finally {
+      setTargetTodosIds([]);
+    }
+  };
+
+  const visibleTodos = useMemo(() => {
+    if (!todos) {
+      return null;
+    }
+
+    switch (filteredBy) {
+      case FilteredBy.All:
+        return todos;
+
+      case FilteredBy.Completed:
+        return todos.filter(todo => todo.completed);
+
+      case FilteredBy.Active:
+        return todos.filter(todo => !todo.completed);
+
+      default:
+        return null;
+    }
+  }, [filteredBy, todos]);
+
+  const setFilter = (value: FilteredBy) => {
+    setFilteredBy(value);
+  };
+
+  const handleInput = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setNewTodoTitle(event.currentTarget.value);
+  };
+
+  const changeTodoStatus = async (todoId: number, completed: boolean) => {
+    setTargetTodosIds(currIds => {
+      return [...currIds, todoId || 0];
+    });
+
+    try {
+      await patchTodos(todoId, { completed });
+
+      setTodos((currTodo) => {
+        if (currTodo) {
+          return currTodo.map(todo => {
+            if (todo.id === todoId) {
+              return { ...todo, completed };
+            }
+
+            return todo;
+          });
+        }
+
+        return null;
+      });
+    } catch (error) {
+      setTypeOfError(Errors.Updating);
+    } finally {
+      setTargetTodosIds([]);
+    }
+  };
+
+  const toggleAll = async () => {
+    const todosForUpdaiting: Promise<unknown>[] = [];
+    const targetTodoIds: number[] = [];
+
+    if (isActive) {
+      todos?.forEach(todo => {
+        if (!todo.completed) {
+          todosForUpdaiting.push(patchTodos(todo.id, { completed: true }));
+          targetTodoIds.push(todo.id);
+        }
+      });
+    } else {
+      todos?.forEach(todo => {
+        todosForUpdaiting.push(patchTodos(todo.id, { completed: false }));
+        targetTodoIds.push(todo.id);
+      });
+    }
```

Comment:
you can simplify sth like this 

```
  todos?.forEach(todo => {
    const completed = isActive ? true : false;
    todosForUpdaiting.push(patchTodos(todo.id, { completed }));
    targetTodoIds.push(todo.id);
  });

```

Reply:
yes, I can. But in this case we will update all todos, not only active. We can try to simplify this part of code, but this simplification will not be very significant.